Source/Dafny/Resolver.cs

```diff
@@ -11709,24 +11709,42 @@ public class ResolveOpts
         ResolveFunctionCallExpr(e, opts);
 
       } else if (expr is ApplyExpr) {
-        var e = (ApplyExpr)expr;
+        var e = (ApplyExpr) expr;
         ResolveExpression(e.Function, opts);
         foreach (var arg in e.Args) {
           ResolveExpression(arg, opts);
         }
+
         // TODO: the following should be replaced by a type-class constraint that constrains the types of e.Function, e.Args[*], and e.Type
         var fnType = e.Function.Type.AsArrowType;
         if (fnType == null) {
-          reporter.Error(MessageSource.Resolver, e.tok, "non-function expression (of type {0}) is called with parameters", e.Function.Type);
+          reporter.Error(MessageSource.Resolver, e.tok,
+            "non-function expression (of type {0}) is called with parameters", e.Function.Type);
         } else if (fnType.Arity != e.Args.Count) {
-          reporter.Error(MessageSource.Resolver, e.tok, "wrong number of arguments to function application (function type '{0}' expects {1}, got {2})", fnType, fnType.Arity, e.Args.Count);
+          reporter.Error(MessageSource.Resolver, e.tok,
+            "wrong number of arguments to function application (function type '{0}' expects {1}, got {2})", fnType,
+            fnType.Arity, e.Args.Count);
         } else {
           for (var i = 0; i < fnType.Arity; i++) {
-            AddAssignableConstraint(e.Args[i].tok, fnType.Args[i], e.Args[i].Type, "type mismatch for argument" + (fnType.Arity == 1 ? "" : " " + i) + " (function expects {0}, got {1})");
+            AddAssignableConstraint(e.Args[i].tok, fnType.Args[i], e.Args[i].Type,
+              "type mismatch for argument" + (fnType.Arity == 1 ? "" : " " + i) + " (function expects {0}, got {1})");
           }
         }
+
         expr.Type = fnType == null ? new InferredTypeProxy() : fnType.Result;
 
+      } else if (expr is SeqConstructionExpr) {
+        var e = (SeqConstructionExpr) expr;
+        var elementType = new InferredTypeProxy();
+        ResolveExpression(e.N, opts);
+        ConstrainToIntegerType(e.N, "sequence construction must use an integer-based expression for the sequence size (got {0})");
+        ResolveExpression(e.Initializer, opts);  // TODO: check type (KRML)
+        var arrowType = new ArrowType(e.tok, builtIns.ArrowTypeDecls[1], new List<Type>() { Type.Int }, elementType);
```

Comment:
I'm still a C# beginner, but is there a nicer way to create arrow types?  Maybe we could have static methods for Function1/2/3 or something.  This is fine, but rather chatty.

Reply:
The C# code that generates Boogie code is very chatty. What you can easily write in a line or two of Boogie, if only you could write Boogie in text, can sometimes take a dozen of lines of C#. I've often wished this were simpler, because it leads to a lot of complexity in the code. There are helper routines for many things, but they are not organized in any disciplined way, so they are typically hard to find. (Maybe sometime we can do a Refactor Fest to improve these things.)

I had missed the TODO comment on line 11741. :O

Reply:
Maybe a pedantic question, but why are we sure `ArrowTypeDecls[1]` is non-null?  I only see `[0]` being created in all cases.